proposals/0243-character-literals.md

```diff
@@ -0,0 +1,107 @@
+ 
+```Swift
+// Modified String literal protocol hierarchy:
+ExpressibleByStringLiteral
+  ↳ ExpressibleByExtendedGraphemeClusterLiteral
+      ↳ ExpressibleByUnicodeScalarLiteral
+          ↳ @_marker ExpressibleBySingleQuotedLiteral
+              ↳ @_marker ExpressibleByASCIILiteral
+```
+This is realised by introducing two new `ExpressibleBy` marker protocols: `ExpressibleBySingleQuotedLiteral` and `ExpressibleByASCIILiteral` which are inserted above the existing `ExpressibleByUnicodeScalarLiteral` in the double quoted literal protocols. As they are prefixed with `@_marker` it is assumed this will not affect the ABI of the existing protocol's witness table layouts. The `ExpressibleBySingleQuotedLiteral` is used only to change the default type of single quoted literals in an expression without type context and the `ExpressibleByASCIILiteral` used to gate the ASCII to integer value conversions. While the inheritance relationship of the ABI-locked `ExpressibleBy` protocols technically entails that `Character` and `Unicode.Scalar` literals can be implicitly promoted to `String` literals, it would be possible in future for the compiler to statically reject such cases at the type checking stage, without affecting ABI, in the interest of untangling the various textual literal forms. As literal delimiters are a purely compile-time construct, and all double-quoted literals currently default to `String`, this will have zero impact on all existing Swift code.
+
+ ## Source compatibility
+
+As the use of the new single quoted syntax is opt-in existing code will continue to compile as before i.e. the proposed implementation is not source breaking. Only where the user has opted to use the new single quoted spelling will the integer conversions be available. This is possible to add a warning and fix-it to prompt the user to move to the new syntax in the course of time. In practice, the Character and Unicode.Scalar types occur do not occur frequently in code so this would not be an arduous migration.
+
+ ## Effect on ABI stability
+
+ Assuming injecting `@_marker` protocols does not alter witness table layout and ABI, this is a purely lexer- and type checker-level change which does not affect the storage or entry points of `Character` and `Unicode.Scalar`. These initializers are an implementation artifact of Swift's protocol-driven literals model, and are meant to be automatically invoked by the compiler. As they are not intended for “public consumption”, we see no reason to continue having the compiler invoke them.
+
+ ## Effect on API resilience
+
+ This is a purely lexer- and type checker-level change which does not affect the API of the standard library.
```

Comment:
it does affect the API of the standard library, because it changes the protocol hierarchy (regardless of `@_marker`ness). it would be more accurate to say that it is *not source-breaking*.